[ApiKeys] Add API key management

clragon opened the pull request:
Adds the ability to have multiple Api keys.

- Api keys now have names, as well as an optional expiration date.
- Each Api key now tracks its last used IP and user agent.
- A usage tracking job that is deduplicated has been added*.
- Api keys are disallowed from accessing Api key routes.
- An index page for Api keys has been added.
- An Api key controller with tests has been implemented.
- The settings page has been updated to point to the new Api key page.
- A migration script to convert the table has been included.

The code in this PR is based on work by @DonovanDMC specifically in https://github.com/FemboyFans/FemboyFans/commit/018eca40e7d549c2e728862d1508be21da4845df.

<sub>\* The usage job can only be triggered once every minute. Any new jobs with the same key during that minute are _rejected_, this means should IPs or user agents differ, they will not be updated within that minute. This seemed like a sensible trade-off to me, rather than continously updating or replacing the job. It would have been possible to track the new parameters in Redis on replacement. Replacing the job on a new request would also pose issues with continues requests not updating anything at all, since the job is always replaced.</sub>

Fixes #736.

<img width="1257" height="476" alt="image" src="https://github.com/user-attachments/assets/05e937b1-79b6-41fe-b754-fb2db923fb5b" />

<img width="1896" height="496" alt="image" src="https://github.com/user-attachments/assets/02df706e-58ed-4c91-9556-46c3b4de0310" />

<img width="362" height="622" alt="image" src="https://github.com/user-attachments/assets/709df8e5-c49e-4e9c-9d59-8cbc5d750f88" />

clragon committed:
[ApiKeys] Add API key management

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

clragon committed:
[ApiKeys] Apply migration

clragon committed:
[ApiKeys] Remove old api key controller

clragon committed:
[ApiKeys] Fix test api keys

clragon committed:
[ApiKeys] Fix tests

clragon committed:
[ApiKeys] Add better expiration, remove editing

clragon committed:
[ApiKeys] Fix review

clragon committed:
[ApiKeys] Improve usage tracking

clragon committed:
[ApiKeys] Improve migration

clragon committed:
[ApiKeys] Reset creation date on regeneration

Sindrake reviewed (changes requested):
The mobile UI looks terrible.

<img width="360" height="800" alt="Image" src="https://github.com/user-attachments/assets/0548d2c9-361b-4173-aceb-ac43fd9ea65a" />

The rest is excellent. Very well done.

Sindrake commented on app/jobs/update_api_key_usage_job.rb:
```diff
@@ -0,0 +1,22 @@
+    api_key = ApiKey.find_by(id: api_key_id)
+    return unless api_key
+
+    api_key.update_usage!(ip_address, user_agent)
```
This will fail for legacy API keys, since they do not have names.
```
ActiveRecord::RecordInvalid: Validation failed: Name can't be blank
```

clragon replied:
The UI has been mobilized.

clragon committed:
[ApiKeys] Make UI responsive

clragon committed:
[ApiKeys] Make name required

clragon committed:
[ApiKeys] Fix tests missing api key name

clragon committed:
[ApiKeys] Fix test utility method usage

clragon committed:
[ApiKeys] Fix migration failing in staged scenarios

clragon committed:
[ApiKeys] Make legacy api keys expire in 6 months

kiranoot commented on app/logical/session_loader.rb:
```diff
@@ -118,7 +118,7 @@ def authenticate_api_key(name, api_key)
     CurrentUser.user = user
     CurrentUser.api_key = api_key_record
-    api_key_record&.update_usage!(@request.remote_ip)
+    UpdateApiKeyUsageJob.set(wait: 1.minute).perform_later(api_key_record.id, @request.remote_ip, @request.user_agent) if api_key_record
```
I believe that this is likely to overload the sidekiq system with jobs. Jobs still have to be distributed to workers and there is overhead related to this.

A better approach would be to use a system like the throttle code uses(Or the throttle code itself). With Redis or some other memory store.

clragon committed:
[ApiKeys] Replace usage job with debounced write

clragon committed:
[ApiKeys] Remove expiration of old keys

Sindrake committed:
Bump migration date

Rolling back 15 migrations when switching between branches is just not practical

Sindrake committed:
Separate API key debouncing

Having debouncing tracking be account-wide may have merit for performance, but the inconsistencies with tracking API key usage it would cause are not worth it, in my opinion.

Sindrake committed:
Add limits to how many API keys a user can have

The limits are a bit conservative right now, but we can bump them up later if need be.

Sindrake committed:
Add expiration notification emails

Sindrake committed:
Minor fixes

Sindrake committed:
Escape the user-agent better
